shared-bindings/vectorio/VectorShape.c

```diff
@@ -0,0 +1,195 @@
+
+#include "shared-module/vectorio/__init__.h"
+#include "shared-bindings/vectorio/VectorShape.h"
+#include "shared-bindings/vectorio/Circle.h"
+#include "shared-bindings/vectorio/Polygon.h"
+#include "shared-bindings/vectorio/Rectangle.h"
+
+#include "shared-bindings/displayio/ColorConverter.h"
+#include "shared-bindings/displayio/Palette.h"
+
+#include <stdint.h>
+
+#include "lib/utils/context_manager_helpers.h"
+
+#include "py/binary.h"
+#include "py/objproperty.h"
+#include "py/objtype.h"
+#include "py/runtime.h"
+#include "supervisor/shared/translate.h"
+
+
+//| .. currentmodule:: vectorio
+//|
+//| :class:`VectorShape` -- Binds a vector shape to a location and pixel color
+//| ==========================================================================
+//|
+//| .. class:: VectorShape( shape, pixel_shader, x=0, y=0)
```

Comment:
TileGrid is well-suited to maximizing bitmap grids.  I tried to use it for `vectorio` but was unable to do so without branching an unreasonable amount.

It does feel similar, yes; but:
* Positioning is a bit different with vector shapes.  `x+width` has nothing to do with the location of the pixels to be drawn in these shapes.
* There is no width and height external to the shapes themselves.  These are not sprite  sheets!
* I'm hoping to make an `mp_shape_protocol` in the future and just put shapes dynamically behind that.
* I'm planning a Rotator that shims in and provides a view of the shape apis rotated about their origin.
* I'm planning an optional antialiasing feature for shapes.  This would be a bit different from a pure 2-color pixel shader bitmap idea.
* I'm a little sad it takes so long to draw a large rectangle and would like to profile to find the slowest parts.  The shape API could be extended to allow shapes to provide batch pixels if their implementations have a fast way to determine batches (rather than 1px at a time).

VectorShape drops a lot of top-down assumptions about shapes' pixels.  It's really difficult to imagine fitting it all into TileGrid but perhaps you've got ideas! =)

</wallOfText>

Reply:
This is fine. :-) Just my initial reaction and clearly you've thought about it way more than me.

> I'm a little sad it takes so long to draw a large rectangle and would like to profile to find the slowest parts. The shape API could be extended to allow shapes to provide batch pixels if their implementations have a fast way to determine batches (rather than 1px at a time).

I thought about doing run length masking and pixel values instead of individual ones. However, I needed to pull myself out of the hole I was in with displayio at the time. I'd love to have you look into performance improvements.